File: AgentCut-ai/app/schemas/plan.py
"""剪辑方案 Pydantic 模型（app.schemas.plan）。

严格对齐 docs/plan-schema.json（跨语言契约，字段名 / 类型 / 必填不可改）：
- 顶层 Plan：schemaVersion / planVersion / projectId / title / source / global / timeline / transitions
- 操作 Operation 为判别联合，用 type 字段的 Literal 区分 speed / crop / subtitle / volume / mute
- 全部模型 extra="forbid"，对应 JSON Schema 的 additionalProperties: false

Python 关键字规避：global -> global_，from -> from_，通过 Field(alias=...) 映射，
序列化时 by_alias=True 输出契约原名字段（global / from）。
"""

from typing import Annotated, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field, model_validator


class TimeRange(BaseModel):
    """源时间区间（浮点秒）。"""

    model_config = ConfigDict(extra="forbid")

    start: float = Field(ge=0)
    end: float = Field(ge=0)


class Source(BaseModel):
    """源视频信息（来自 PyAV 探测 / ffprobe）。"""

    model_config = ConfigDict(extra="forbid")

    assetId: str
    url: str
    duration: float = Field(ge=0)
    fps: float = Field(gt=0)
    width: int = Field(gt=0)
    height: int = Field(gt=0)


class OutputConfig(BaseModel):
    """成片输出配置。"""

    model_config = ConfigDict(extra="forbid")

    width: int = Field(gt=0)
    height: int = Field(gt=0)
    fps: float = Field(gt=0)
    codec: str = Field(default="libx264")
    bitrate: Optional[str] = None


class Bgm(BaseModel):
    """背景音乐（全局级）。"""

    model_config = ConfigDict(extra="forbid")

    url: str
    volume: float = Field(default=0.3, ge=0, le=1)
    loop: bool = Field(default=True)


class SubtitleStyle(BaseModel):
    """字幕样式。"""

    model_config = ConfigDict(extra="forbid")

    fontSize: Optional[int] = Field(default=None, ge=1)
    color: Optional[str] = Field(default=None, pattern=r"^#[0-9a-fA-F]{6}$")
    position: Optional[Literal["top", "bottom", "center"]] = None


class Global(BaseModel):
    """全局设置。"""

    model_config = ConfigDict(extra="forbid")

    output: OutputConfig
    bgm: Optional[Bgm] = None
    subtitleStyle: Optional[SubtitleStyle] = None


# ---------------------------------------------------------------------------
# 原子操作（判别联合，按 type 分派）
# ---------------------------------------------------------------------------


class OpSpeed(BaseModel):
    """变速：rate 为速度倍率（>0）。"""

    model_config = ConfigDict(extra="forbid")

    type: Literal["speed"]
    rate: float = Field(gt=0)


class OpCrop(BaseModel):
    """裁切：源像素坐标下的裁剪窗口。"""

    model_config = ConfigDict(extra="forbid")

    type: Literal["crop"]
    x: float = Field(ge=0)
    y: float = Field(ge=0)
    width: float = Field(gt=0)
    height: float = Field(gt=0)


class OpSubtitle(BaseModel):
    """字幕：text 与段内时间范围（start/end 相对片段起点）。"""

    model_config = ConfigDict(extra="forbid")

    type: Literal["subtitle"]
    text: str
    start: float = Field(ge=0)
    end: float = Field(ge=0)


class OpVolume(BaseModel):
    """音量调节：volume 0~1。"""

    model_config = ConfigDict(extra="forbid")

    type: Literal["volume"]
    volume: float = Field(ge=0, le=1)


class OpMute(BaseModel):
    """静音。"""

    model_config = ConfigDict(extra="forbid")

    type: Literal["mute"]


# 判别联合：与 plan-schema.json 的 operation.oneOf 一致
Operation = Annotated[
    Union[OpSpeed, OpCrop, OpSubtitle, OpVolume, OpMute],
    Field(discriminator="type"),
]


class Segment(BaseModel):
    """时间线片段（顺序即成片顺序；keep=false 表示剪掉）。"""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(pattern=r"^[A-Za-z0-9_-]+$")
    keep: bool
    sourceRange: TimeRange
    operations: list[Operation] = Field(default_factory=list)


class Transition(BaseModel):
    """片段间转场。"""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    from_: str = Field(alias="from")
    to: str
    type: Literal["fade", "none"]
    duration: float = Field(default=0.5, ge=0)


class Plan(BaseModel):
    """剪辑方案（跨语言契约，严格对齐 docs/plan-schema.json）。"""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schemaVersion: Literal["1.0"] = "1.0"
    planVersion: int = Field(ge=1)
    projectId: str
    title: Optional[str] = None
    source: Source
    global_: Global = Field(alias="global")
    timeline: list[Segment] = Field(min_length=1)
    transitions: list[Transition] = Field(default_factory=list)

    @model_validator(mode="after")
    def _validate_references(self) -> "Plan":
        """转场引用的片段必须存在于 timeline（保证方案可执行）。"""
        ids = {seg.id for seg in self.timeline}
        for tr in self.transitions:
            if tr.from_ not in ids or tr.to not in ids:
                raise ValueError(f"transition 引用不存在的片段: {tr.from_} -> {tr.to}")
        return self

    def to_contract_dict(self) -> dict:
        """序列化为与 plan-schema.json 对齐的 dict。

        - by_alias：输出契约原名字段（global / from）
        - exclude_none：可选字段缺省时整体省略而非输出 null
          （plan-schema 中可选属性均为纯类型，不允许 null，如 title / bgm / subtitleStyle / bitrate）
        """
        return self.model_dump(mode="json", by_alias=True, exclude_none=True)

    def to_contract_json(self) -> str:
        """序列化为与 plan-schema.json 对齐的 JSON 字符串。"""
        return self.model_dump_json(by_alias=True, exclude_none=True)

File: AgentCut-front/src/types/plan.ts
/**
 * 剪辑方案 TypeScript 类型定义
 *
 * 唯一契约源：E:\java\AgentCut\docs\plan-schema.json
 * 后端 GET /api/v1/plans/{projectId} 返回的 JSON 即对齐此类型。
 * 核心原则：方案文档是单一事实来源 —— Agent 生成、人编辑、引擎执行、系统存档共用同一份文档。
 */

/** 源时间区间（绝对浮点秒） */
export interface TimeRange {
  start: number;
  end: number;
}

/** 输出配置 */
export interface OutputConfig {
  width: number;
  height: number;
  fps: number;
  codec?: string;
  bitrate?: string;
}

/** 背景音乐（global 级） */
export interface Bgm {
  url: string;
  volume?: number;
  loop?: boolean;
}

/** 字幕样式 */
export interface SubtitleStyle {
  fontSize?: number;
  color?: string;
  position?: 'top' | 'bottom' | 'center';
}

/** 全局设置 */
export interface Global {
  output: OutputConfig;
  bgm?: Bgm;
  subtitleStyle?: SubtitleStyle;
}

/** 源视频信息（来自 ffprobe） */
export interface Source {
  assetId: string;
  url: string;
  duration: number;
  fps: number;
  width: number;
  height: number;
}

/* ==================== 操作（判别联合） ==================== */

/** 操作类型名 */
export type OperationType = 'speed' | 'crop' | 'subtitle' | 'volume' | 'mute';

/** 原子剪辑操作：以 type 字段区分，判别联合（discriminated union） */
export type Operation = SpeedOp | CropOp | SubtitleOp | VolumeOp | MuteOp;

/** 变速 */
export interface SpeedOp {
  type: 'speed';
  rate: number;
}

/** 裁切 */
export interface CropOp {
  type: 'crop';
  x: number;
  y: number;
  width: number;
  height: number;
}

/** 字幕 */
export interface SubtitleOp {
  type: 'subtitle';
  text: string;
  start: number;
  end: number;
}

/** 音量 */
export interface VolumeOp {
  type: 'volume';
  volume: number;
}

/** 静音（无参数） */
export interface MuteOp {
  type: 'mute';
}

/* ==================== 片段 / 转场 / 方案 ==================== */

/** 时间线片段（顺序即最终成片顺序） */
export interface Segment {
  id: string;
  keep: boolean;
  sourceRange: TimeRange;
  operations?: Operation[];
}

/** 转场 */
export interface Transition {
  from: string;
  to: string;
  type: 'fade' | 'none';
  duration?: number;
}

/** 剪辑方案（顶层文档） */
export interface Plan {
  schemaVersion: '1.0';
  planVersion: number;
  projectId: string;
  title?: string;
  source: Source;
  global: Global;
  timeline: Segment[];
  transitions?: Transition[];
}

/* ==================== 辅助：编辑器用 ==================== */

/** 操作类型中文文案 */
export const OPERATION_LABELS: Record<OperationType, string> = {
  speed: '变速',
  crop: '裁切',
  subtitle: '字幕',
  volume: '音量',
  mute: '静音',
};

/** 新增操作时的参数模板 */
export function createOperation(type: OperationType): Operation {
  switch (type) {
    case 'speed':
      return { type: 'speed', rate: 1.0 };
    case 'crop':
      return { type: 'crop', x: 0, y: 0, width: 1080, height: 1920 };
    case 'subtitle':
      return { type: 'subtitle', text: '', start: 0, end: 0 };
    case 'volume':
      return { type: 'volume', volume: 1.0 };
    case 'mute':
      return { type: 'mute' };
  }
}

/** 空方案工厂（便于前端在无后端时构造演示数据） */
export function createEmptyPlan(projectId: string): Plan {
  return {
    schemaVersion: '1.0',
    planVersion: 1,
    projectId,
    source: {
      assetId: '',
      url: '',
      duration: 0,
      fps: 30,
      width: 1920,
      height: 1080,
    },
    global: {
      output: { width: 1080, height: 1920, fps: 30 },
    },
    timeline: [],
  };
}
